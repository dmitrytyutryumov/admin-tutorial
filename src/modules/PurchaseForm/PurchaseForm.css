.purchase-form {
  display: grid;
  justify-items: center;
  align-items: center;
  gap: 1rem;

  background-color: rgb(248, 248, 248);
  border: rgb(135, 134, 134) solid 1px;
  border-radius: 0.3rem;
}

.purchase-form__title {
  font-family: 'Courier New', Courier, monospace;
  font-size: 2rem;
  line-height: 1.1;

  font-weight: 500;
}

.purchase-form__body {
  display: grid;
  justify-items: start;
  gap: 1rem;
}

.purchase-form__btn {
  width: 100%;
  min-height: 2rem;
}
