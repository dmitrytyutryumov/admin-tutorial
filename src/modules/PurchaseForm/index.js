export { PurchaseForm } from './PurchaseForm'
import { fetchPurchase } from '../../api/Purchases'

export async function loader({ params }) {
  return await fetchPurchase(params.id)
}
