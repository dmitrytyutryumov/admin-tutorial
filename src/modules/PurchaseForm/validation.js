import * as Yup from 'yup'

export const validationSchema = Yup.object({
  jan: Yup.number().required(),
  feb: Yup.number().required(),
  mar: Yup.number().required(),
  apr: Yup.number().required(),
  may: Yup.number().required(),
  jun: Yup.number().required(),
  jul: Yup.number().required(),
  aug: Yup.number().required(),
  sep: Yup.number().required(),
  oct: Yup.number().required(),
  nov: Yup.number().required(),
  dec: Yup.number().required(),
  language: Yup.string().required(),
  name: Yup.string().required(),
  gameName: Yup.string().required(),
  rating: Yup.number().required(),
  bankBalance: Yup.number().required(),
  bought: Yup.bool().required(),
  totalWinnings: Yup.number().required(),
})
