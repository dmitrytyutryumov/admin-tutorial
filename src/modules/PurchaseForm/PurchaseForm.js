import { useLoaderData, useNavigate } from 'react-router-dom'
import { Formik, Form } from 'formik'
import { updatePurchase } from '../../api/Purchases'
import { InputField } from './components/InputField'

import './PurchaseForm.css'

export function PurchaseForm() {
  const { data } = useLoaderData()
  const navigate = useNavigate()

  const onSubmit = async (values) => {
    await updatePurchase(data.id, values)
    navigate('/')
  }

  const fields = Object.keys(data).map((name) => {
    if (name === 'id') {
      return
    }
    return <InputField name={name} key={name} />
  })

  return (
    <Formik initialValues={{ ...data }} onSubmit={onSubmit}>
      <Form className="purchase-form">
        <h1 className="purchase-form__title">Purchase Details</h1>
        <div className="purchase-form__body">
          {fields}
          <button className="btn purchase-form__btn" type="submit">
            Submit
          </button>
        </div>
      </Form>
    </Formik>
  )
}
