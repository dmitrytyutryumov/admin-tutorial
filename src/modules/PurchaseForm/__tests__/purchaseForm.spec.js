import { render } from '@testing-library/react'
import { useField } from 'formik'
import { useLoaderData } from 'react-router-dom'
import { PurchaseForm } from '../PurchaseForm'

describe('test purchase form', () => {
  beforeEach(() => {
    useField.mockImplementation(() => [
      { value: 1, onChange: jest.fn(), onBlue: jest.fn() },
      { error: null, touched: false },
      { setTouched: jest.fn() },
    ])
  })
  it('test purchase form', () => {
    // Arrange
    useLoaderData.mockImplementation(() => {
      return {
        data: {
          id: 1,
          string: 'field1',
          number: 1,
          boolean: true,
          array: [1, 2],
        },
      }
    })
    // Act
    const { asFragment } = render(<PurchaseForm />)
    // Assert
    expect(asFragment()).toMatchSnapshot()
  })
})
