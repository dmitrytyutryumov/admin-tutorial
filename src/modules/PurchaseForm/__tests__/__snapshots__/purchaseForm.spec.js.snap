// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test purchase form test purchase form 1`] = `
<DocumentFragment>
  <form
    action="#"
    class="purchase-form"
  >
    <h1
      class="purchase-form__title"
    >
      Purchase Details
    </h1>
    <div
      class="purchase-form__body"
    >
      <div
        class="form-input-wrapper"
      >
        <label
          class="input-label "
          for="string"
        >
          String
        </label>
        <input
          checked=""
          class="text-input "
          id="string"
          name="string"
          type="number"
          value="1"
        />
      </div>
      <div
        class="form-input-wrapper"
      >
        <label
          class="input-label "
          for="number"
        >
          Number
        </label>
        <input
          checked=""
          class="text-input "
          id="number"
          name="number"
          type="number"
          value="1"
        />
      </div>
      <div
        class="form-input-wrapper"
      >
        <label
          class="input-label "
          for="boolean"
        >
          Boolean
        </label>
        <input
          checked=""
          class="text-input "
          id="boolean"
          name="boolean"
          type="number"
          value="1"
        />
      </div>
      <div
        class="form-input-wrapper"
      >
        <label
          class="input-label "
          for="array"
        >
          Array
        </label>
        <input
          checked=""
          class="text-input "
          id="array"
          name="array"
          type="number"
          value="1"
        />
      </div>
      <button
        class="btn purchase-form__btn"
        type="submit"
      >
        Submit
      </button>
    </div>
  </form>
</DocumentFragment>
`;
